fix(contact-request): fix various contact request issues

jrainville opened the pull request:
Fixes #4420

-Makes the contact request appear "in real time" instead of having to reload
-Creates the 1-1 chat with the new contact when added (the qml was already fixed, but it was fixed by removing the faulty feature instead of fixing it)
- The crash that was originally reported by Pascal had been fixed a while ago. The crash reported by Iuri, I wasn't able to reproduce and I could add and remove contact requests without issues.

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 86f4a2cd | [#1](https://ci.status.im/job/status-desktop/job/branches/job/macos/job/PR-4434/1/) | 2022-01-12 21:20:05 | ~17 min | `macos` | [:package:`dmg`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Desktop-220112-210223-86f4a2-pr4434.dmg) |

anastasiyaig commented:
@jrainville the PR crashes on chatting with a contact

**Steps:**

1. on desktop join a public chat and send some messages
2. on mobile join the same public chat and send some messages
3. on desktop, add the mobile to contacts
4. on desktop, go to settings -> contacts -> click the mobile u just added to open 1x1 chat (since profile -> send message is not working yet)
5. in this 1x1 chat with mobile, send some messages
6. on mobile, open activity center and accept the contact request
7. on mobile, in 1x1 chat send some messages

As result, the desktop app crashes. I checked in base_bc and it is not crashing there, however opening 1x1 chat is much more harder there (and maybe it makes a difference). Could u please take a look?

```
INF 2022-01-13 09:52:31.676+03:00 shutting down...                           topics="status-app" tid=545533 file=nim_status_client.nim:121
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/nim_status_client.nim(146) nim_status_client
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/nim_status_client.nim(143) mainProc
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/vendor/nimqml/src/nimqml/private/qguiapplication.nim(29) exec
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/vendor/nimqml/src/nimqml/private/qobject.nim(49) qobjectCallback
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/vendor/nimqml/src/nimqml/private/qobject.nim(34) onSlotCalled
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/vendor/nimbus-build-system/vendor/Nim/lib/core/macros.nim(566) onSlotCalled
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/app/core/signals/signals_manager.nim(56) receiveSignal
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/app/core/signals/signals_manager.nim(53) processSignal
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/app/core/eventemitter.nim(46) emit
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/app_service/service/contacts/service.nim(122) :anonymous
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/app/core/eventemitter.nim(46) emit
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/app/modules/main/profile_section/contacts/controller.nim(49) :anonymous
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/app/modules/main/profile_section/contacts/io_interface.nim(67) contactUpdated
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/app/modules/main/profile_section/contacts/module.nim(103) contactUpdated
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/src/app/modules/shared_models/contacts_model.nim(132) updateItem
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/vendor/nimqml/src/nimqml/private/qabstractitemmodel.nim(237) dataChanged
/Users/jenkins/workspace/s-desktop_branches_macos_PR-4434/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(49) sysFatal
Error: unhandled exception: index out of bounds, the container is empty [IndexError]
```

saledjenic commented on src/app/modules/main/chat_section/module.nim:
```diff
@@ -399,4 +400,9 @@ method onContactBlocked*(self: Module, publicKey: string) =
   let (chatName, chatImage, isIdenticon) = self.controller.getOneToOneChatNameAndImage(publicKey)
+  let contact = self.controller.getContactById(publicKey)
+  if (contact.hasAddedUs and not contact.added):
+    let item = self.createItemFromPublicKey(publicKey)
```
@jrainville I know that what I am going to say is not part of the changes made here, but I noticed that now, so in the method `createItemFromPublicKey` we have two calls to the contacts service:
```
proc createItemFromPublicKey(self: Module, publicKey: string): contacts_item.Item =
  let contact =  self.controller.getContact(publicKey)
  let (name, image, isIdenticon) = self.controller.getContactNameAndImage(contact.id)
  
  return contacts_item.initItem(contact.id, name, image, isIdenticon, contact.isContact(), contact.isBlocked(), contact.requestReceived())
```
but instead we can make it simpler, since we already have convenient method in contacts service named `getContactDetails`, so if you want you may add it to the controller and use in `createItemFromPublicKey` like this:
```
proc createItemFromPublicKey(self: Module, publicKey: string): contacts_item.Item =
  let contactDetails = self.controller.getContactDetails(publicKey)
  
  return contacts_item.initItem(contactDetails.details.id, contactDetails.displayName, contactDetails.icon, contactDetails.isIdenticon, contactDetails.details.isContact(), contactDetails.details.isBlocked(), contactDetails.details.requestReceived())
```

saledjenic commented on src/app/modules/main/chat_section/module.nim:
```diff
@@ -399,4 +400,9 @@ method onContactBlocked*(self: Module, publicKey: string) =
 method onContactDetailsUpdated*(self: Module, publicKey: string) =
   let (chatName, chatImage, isIdenticon) = self.controller.getOneToOneChatNameAndImage(publicKey)
+  let contact = self.controller.getContactById(publicKey)
+  if (contact.hasAddedUs and not contact.added):
```
@jrainville we need to check here if contact is not blocked as well, so please updated this condition to (like it's in `initContactRequestsModel` proc):
```
if(contact.requestReceived() and not contact.isContact() and not contact.isBlocked()):
```

saledjenic commented on src/app/modules/shared_models/contacts_model.nim:
```diff
@@ -86,6 +92,10 @@ QtObject:
 
   proc addItem*(self: Model, item: Item) =
+    let ind = self.findIndexByPubKey(item.pubKey)
+    if(ind > -1):
```
@jrainville if the point of this check is to add item only in case it was not added before, then that should be a case for a business logic part... and there is already a method `containsItemWithPubKey` which returns `true` if an item with a public key is already in the model. So please revert changes done here and update `onContactDetailsUpdated` method in the `module.nim`, it should look something like this:
```
method onContactDetailsUpdated*(self: Module, publicKey: string) =
  let contact = self.controller.getContactById(publicKey)
  if(contact.requestReceived() and 
    not contact.isContact() and 
    not contact.isBlocked() and
    not self.view.contactRequestsModel().containsItemWithPubKey(publicKey)):
      let item = self.createItemFromPublicKey(publicKey)
      self.view.contactRequestsModel().addItem(item)

  let (chatName, chatImage, isIdenticon) = self.controller.getOneToOneChatNameAndImage(publicKey)
  self.view.chatsModel().updateItemDetails(publicKey, chatName, chatImage, isIdenticon)
```

saledjenic reviewed (changes requested):
@jrainville all good here, I added some comments which are more like minor improvements, just to keep things clear.

I didn't check a crash @anastasiyaig mentioned, if I may help somehow about it, let me know.

jrainville commented:
I rebased and fixed the review comments @saledjenic , you can review again.

@anastasiyaig I fixed the crash. As for the contact removal, I think maybe Sale's PR (https://github.com/status-im/status-desktop/pull/4452)  might fix it. If it doesn't, I think we should open it as a separate issue/PR, since this PR is more about contact requests.

Finally, the last warning you are seeing is something that happens in master as well. It's the Delivered signal that we haven't implemented yet. The signal manager is just warning that it received a signal that it doesn't know about. It's gonna be implemented someday I guess. It's a whole new feature.

saledjenic commented on src/app/modules/shared_models/contacts_model.nim:
```diff
@@ -124,7 +125,16 @@ QtObject:
     self.items[ind] = item
 
     let index = self.createIndex(ind, 0, nil)
-    self.dataChanged(index, index, @[]) # all roles
```
@jrainville using empty roles array doesn't update all roles for the item? That's weird since see [documentation](https://doc.qt.io/qt-5/qabstractitemmodel.html#dataChanged).

jrainville replied:
I assume NimQml doesn't handle it correctly

jrainville commented:
Merging now because the changes are needed to fix the crash for other devs
